File: 0690_employee_importance/solution.go
// LeetCode 0690 - Employee Importance
// https://leetcode.com/problems/employee-importance/

type Employee struct {
	Id           int
	Importance   int
	Subordinates []int
}

func getImportance(employees []*Employee, id int) int {
	table := map[int]*Employee{}
	for _, emp := range employees {
		table[emp.Id] = emp
	}
	var dfs func(eid int) int
	dfs = func(eid int) int {
		emp := table[eid]
		total := emp.Importance
		for _, sub := range emp.Subordinates {
			total += dfs(sub)
		}
		return total
	}
	return dfs(id)
}

File: 0690_employee_importance/Solution.scala
// LeetCode 0690 - Employee Importance
// https://leetcode.com/problems/employee-importance/

class Employee(var id: Int = 0, var importance: Int = 0, var subordinates: List[Int] = Nil)

object Solution {
  def getImportance(employees: List[Employee], id: Int): Int = {
    val table = scala.collection.mutable.HashMap.empty[Int, Employee]
    for (emp <- employees) table(emp.id) = emp
    def dfs(eid: Int): Int = {
      val emp = table(eid)
      var total = emp.importance
      for (sub <- emp.subordinates) total += dfs(sub)
      total
    }
    dfs(id)
  }
}
